File: src/main/java/slavsquatsuperstar/mayonez/renderer/DebugDraw.java
package slavsquatsuperstar.mayonez.renderer;

import slavsquatsuperstar.mayonez.Game;
import slavsquatsuperstar.mayonez.Vector2;
import slavsquatsuperstar.mayonez.physics2d.primitives.*;

import java.awt.*;
import java.awt.geom.Ellipse2D;
import java.awt.geom.Line2D;
import java.awt.geom.Rectangle2D;
import java.util.ArrayList;
import java.util.List;

/**
 * Draws colliders and mathematical objects onto the screen. All methods use world coordinates, and shapes are centered
 * around the collider's position.
 */
public class DebugDraw {

    private static final int STROKE_SIZE = 2;
    private static final Stroke stroke = new BasicStroke(STROKE_SIZE);
    private static final List<Renderable> shapes = new ArrayList<>(); // map to color?

    // Points

    /**
     * Draws a point onto the screen
     *
     * @param position where the point is located, in world coordinates
     * @param color    the color to use
     */
    public static void drawPoint(Vector2 position, Color color) {
        float radiusPx = STROKE_SIZE; // Fill a circle with radius "STROKE_SIZE"
        shapes.add(g2 -> {
            g2.setColor(color);
            g2.fill(new Ellipse2D.Float(toScreen(position.x) - radiusPx, toScreen(position.y) - radiusPx, radiusPx * 2, radiusPx * 2));
        });
    }

    // Lines

    /**
     * Draws a line segment onto the screen.
     *
     * @param start the segment's starting point, in world coordinates
     * @param end   the segment's ending point in, world coordinates
     * @param color the color to use
     */
    public static void drawLine(Vector2 start, Vector2 end, Color color) {
        Vector2 startPx = toScreen(start);
        Vector2 endPx = toScreen(end);
        shapes.add(g2 -> {
            g2.setColor(color);
            g2.draw(new Line2D.Float(startPx.x, startPx.y, endPx.x, endPx.y));
        });
    }

    /**
     * Draws a vector onto the screen.
     *
     * @param origin    the vector's starting point, in world coordinates
     * @param direction how far away the vector's end point is, in world coordinates
     * @param color     the color to use
     */
    public static void drawVector(Vector2 origin, Vector2 direction, Color color) {
        drawLine(origin, origin.add(direction), color);
//        drawPoint(origin.add(direction), color); // draw the "arrowhead"
    }

    // Shapes (Outline)

    /**
     * Draws a shape onto the screen.
     *
     * @param shape a {@link Collider2D} instance
     * @param color the color to use
     */
    public static void drawShape(Collider2D shape, Color color) {
        if (shape instanceof CircleCollider)
            drawCircle((CircleCollider) shape, color);
        else if (shape instanceof AlignedBoxCollider2D)
            drawAABB((AlignedBoxCollider2D) shape, color);
        else if (shape instanceof BoxCollider2D)
            drawBox((BoxCollider2D) shape, color);
        else if (shape instanceof Edge2D)
            drawLine(((Edge2D) shape).start, ((Edge2D) shape).end, color);
    }

    private static void drawCircle(CircleCollider circle, Color color) {
        Vector2 minPx = toScreen(circle.min());
        float diameterPx = toScreen(circle.radius * 2);
        shapes.add(g2 -> {
            g2.setColor(color);
            g2.draw(new Ellipse2D.Float(minPx.x, minPx.y, diameterPx, diameterPx));
        });
    }

    private static void drawAABB(AlignedBoxCollider2D aabb, Color color) {
        Vector2 minPx = toScreen(aabb.min());
        shapes.add(g2 -> {
            g2.setColor(color);
            g2.draw(new Rectangle2D.Float(minPx.x, minPx.y, toScreen(aabb.width()), toScreen(aabb.height())));
        });
    }

    private static void drawBox(BoxCollider2D box, Color color) {
        Polygon obb = new Polygon();
        for (Vector2 point : box.getVertices())
            obb.addPoint(Math.round(toScreen(point.x)), Math.round(toScreen(point.y)));
        shapes.add(g2 -> {
            g2.setColor(color);
            g2.drawPolygon(obb);
        });
    }

    // Shapes (Filled)

    /**
     * Fills in a shape onto the screen.
     *
     * @param shape a {@link Collider2D} instance
     * @param color the color to use
     */
    public static void fillShape(Collider2D shape, Color color) {
        if (shape instanceof CircleCollider)
            fillCircle((CircleCollider) shape, color);
        else if (shape instanceof AlignedBoxCollider2D)
            fillAABB((AlignedBoxCollider2D) shape, color);
        else if (shape instanceof BoxCollider2D)
            fillBox((BoxCollider2D) shape, color);
        else if (shape instanceof Edge2D)
            drawLine(((Edge2D) shape).start, ((Edge2D) shape).end, color);
    }

    private static void fillCircle(CircleCollider circle, Color color) {
        Vector2 minPx = toScreen(circle.min());
        float diameterPx = toScreen(circle.radius * 2);
        shapes.add(g2 -> {
            g2.setColor(color);
            g2.fill(new Ellipse2D.Float(minPx.x, minPx.y, diameterPx, diameterPx));
        });
    }

    private static void fillAABB(AlignedBoxCollider2D aabb, Color color) {
        Vector2 minPx = toScreen(aabb.min());
        shapes.add(g2 -> {
            g2.setColor(color);
            g2.fill(new Rectangle2D.Float(minPx.x, minPx.y, toScreen(aabb.width()), toScreen(aabb.height())));
        });
    }

    private static void fillBox(BoxCollider2D box, Color color) {
        Polygon obb = new Polygon();
        for (Vector2 point : box.getVertices())
            obb.addPoint(Math.round(toScreen(point.x)), Math.round(toScreen(point.y)));
        shapes.add(g2 -> {
            g2.setColor(color);
            g2.fillPolygon(obb);
        });
    }

    /**
     * Converts a world coordinate to screen coordinates.
     *
     * @param world the location of something in the world along an axis
     * @return the corresponding screen pixel
     */
    private static float toScreen(float world) {
        return world * Game.currentScene().getCellSize();
    }

    /**
     * Converts a world position to screen coordinates.
     *
     * @param world the location of something in the world along both axes
     * @return the corresponding screen pixels
     */
    private static Vector2 toScreen(Vector2 world) {
        return world.mul(Game.currentScene().getCellSize());
    }

    public void render(Graphics2D g2) {
        if (!shapes.isEmpty()) {
            g2.setStroke(stroke);
            shapes.forEach(s -> s.draw(g2));
            shapes.clear();
        }
    }

}

File: src/main/kotlin/slavsquatsuperstar/mayonez/renderer/DebugDraw.kt
package slavsquatsuperstar.mayonez.renderer

import slavsquatsuperstar.math.Vec2
import slavsquatsuperstar.mayonez.Game
import slavsquatsuperstar.mayonez.physics2d.colliders.AlignedBoxCollider2D
import slavsquatsuperstar.mayonez.physics2d.colliders.CircleCollider
import slavsquatsuperstar.mayonez.physics2d.colliders.Collider2D
import slavsquatsuperstar.mayonez.physics2d.colliders.PolygonCollider2D
import java.awt.*
import java.awt.geom.Ellipse2D
import java.awt.geom.Line2D
import java.awt.geom.Rectangle2D
import java.util.function.Consumer
import kotlin.math.roundToInt

/**
 * Draws colliders and mathematical objects onto the screen. All methods use world coordinates, and shapes are centered
 * around the collider's position.
 */
object DebugDraw {

    private const val STROKE_SIZE = 2
    private val stroke: Stroke = BasicStroke(STROKE_SIZE.toFloat())
    private val shapes: MutableList<Renderable> = ArrayList() // map to color?

    // Game Loop Method
    fun render(g2: Graphics2D) {
        if (shapes.isNotEmpty()) {
            g2.stroke = stroke
            shapes.forEach(Consumer { s: Renderable -> s.draw(g2) })
            shapes.clear()
        }
    }

    // Draw Point
    /**
     * Draws a point onto the screen
     *
     * @param position where the point is located, in world coordinates
     * @param color    the color to use
     */
    @JvmStatic
    fun drawPoint(position: Vec2, color: Color?) {
        val radiusPx = STROKE_SIZE.toFloat() // Fill a circle with radius "STROKE_SIZE" in pixels
        shapes.add(Renderable { g2: Graphics2D ->
            g2.color = color
            g2.fill(
                Ellipse2D.Float(
                    position.x.toScreen() - radiusPx, position.y.toScreen() - radiusPx,
                    radiusPx * 2, radiusPx * 2
                )
            )
        })
    }

    // Draw Lines

    /**
     * Draws a line segment onto the screen.
     *
     * @param start the segment's starting point, in world coordinates
     * @param end   the segment's ending point in, world coordinates
     * @param color the color to use
     */
    @JvmStatic
    fun drawLine(start: Vec2, end: Vec2, color: Color?) {
        val startPx = start.toScreen()
        val endPx = end.toScreen()
        shapes.add(Renderable { g2: Graphics2D ->
            g2.color = color
            g2.draw(Line2D.Float(startPx.x, startPx.y, endPx.x, endPx.y))
        })
    }

    /**
     * Draws a vector onto the screen.
     *
     * @param origin    the vector's starting point, in world coordinates
     * @param direction how far away the vector's end point is, in world coordinates
     * @param color     the color to use
     */
    @JvmStatic
    fun drawVector(origin: Vec2, direction: Vec2?, color: Color?) {
        drawLine(origin, origin.add(direction!!), color)
        //        drawPoint(origin.add(direction), color); // draw the "arrowhead"
    }

    // Draw Shapes

    /**
     * Draws a shape onto the screen.
     *
     * @param shape a [Collider2D] instance
     * @param color the color to use
     */
    @JvmStatic
    fun drawShape(shape: Collider2D?, color: Color) {
        when (shape) {
            is CircleCollider -> drawCircle(shape, color)
            is AlignedBoxCollider2D -> drawAABB(shape, color)
            is PolygonCollider2D -> drawPolygon(shape, color)
        }
    }

    private fun drawCircle(circle: CircleCollider, color: Color) {
        val minPx = (circle.center() - Vec2(circle.radius(), circle.radius())).toScreen()
        val diameterPx = (circle.radius() * 2).toScreen()
        shapes.add(Renderable { g2: Graphics2D ->
            g2.color = color
            g2.draw(Ellipse2D.Float(minPx.x, minPx.y, diameterPx, diameterPx))
        })
    }

    private fun drawAABB(aabb: AlignedBoxCollider2D, color: Color) {
        val minPx = aabb.min().toScreen()
        shapes.add(Renderable { g2: Graphics2D ->
            g2.color = color
            g2.draw(Rectangle2D.Float(minPx.x, minPx.y, aabb.width().toScreen(), aabb.height().toScreen()))
        })
    }

    private fun drawPolygon(polygon: PolygonCollider2D, color: Color) {
        val shape = Polygon()
        for (point in polygon.getVertices())
            shape.addPoint(point.x.toScreen().roundToInt(), point.y.toScreen().roundToInt())
        shapes.add(Renderable { g2: Graphics2D ->
            g2.color = color
            g2.drawPolygon(shape)
        })
    }

    // Fill Shapes

    /**
     * Fills in a shape onto the screen.
     *
     * @param shape a [Collider2D] instance
     * @param color the color to use
     */
    @JvmStatic
    fun fillShape(shape: Collider2D?, color: Color) {
        when (shape) {
            is CircleCollider -> fillCircle(shape, color)
            is AlignedBoxCollider2D -> fillAABB(shape, color)
            is PolygonCollider2D -> fillPolygon(shape, color)
        }
    }

    private fun fillCircle(circle: CircleCollider, color: Color) {
        val minPx = (circle.center() - Vec2(circle.radius(), circle.radius())).toScreen()
        val diameterPx = (circle.radius() * 2).toScreen()
        shapes.add(Renderable { g2: Graphics2D ->
            g2.color = color
            g2.fill(Ellipse2D.Float(minPx.x, minPx.y, diameterPx, diameterPx))
        })
    }

    private fun fillAABB(aabb: AlignedBoxCollider2D, color: Color) {
        val minPx = aabb.min().toScreen()
        shapes.add(Renderable { g2: Graphics2D ->
            g2.color = color
            g2.fill(Rectangle2D.Float(minPx.x, minPx.y, aabb.width().toScreen(), aabb.height().toScreen()))
        })
    }

    private fun fillPolygon(box: PolygonCollider2D, color: Color) {
        val shape = Polygon()
        for (point in box.getVertices())
            shape.addPoint(point.x.toScreen().roundToInt(), point.y.toScreen().roundToInt())
        shapes.add(Renderable { g2: Graphics2D ->
            g2.color = color
            g2.fillPolygon(shape)
        })
    }

    // Helper Methods

    /**
     * Converts a coordinate from world to screen units.
     *
     * @return the corresponding screen pixel
     */
    private fun Float.toScreen(): Float = this * Game.currentScene().cellSize

    /**
     * Converts a pair of coordinates from world to screen units.
     *
     * @return the corresponding screen pixels
     */
    private fun Vec2.toScreen(): Vec2 = this * Game.currentScene().cellSize.toFloat()
}